atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -2509,38 +2563,80 @@ private void logCommitLockTenureExceeded(
                 stackTraceSnapshot);
     }
 
-    protected Set<LockDescriptor> getLocksForWrites() {
-        Set<LockDescriptor> result = new HashSet<>();
+    protected LocksAndMetadata getLocksAndMetadataForWrites() {
+        Set<LockDescriptor> lockDescriptors = new HashSet<>();
+        Map<LockDescriptor, ChangeMetadata> lockDescriptorToChangeMetadata = new HashMap<>();
```

Comment:
We should take the opportunity and make these immutable! By using a map builder instead.

Reply:
We won't be able to get size check, but I think we should either then have `collectCellLocks` itself return a map, or an integer indicating the number of entries it added

Reply:
Yeah why not